addrs/provider.go

```diff
@@ -0,0 +1,47 @@
+package addrs
+
+import svchost "github.com/hashicorp/terraform-svchost"
+
+// Provider encapsulates a single provider type. In the future this will be
+// extended to include additional fields including Namespace and SourceHost
+type Provider struct {
+	Type      string
+	Namespace string
+	Hostname  svchost.Hostname
+}
+
+// String returns an FQN string, indended for use in output.
+func (pt Provider) String() string {
+	return pt.Hostname.ForDisplay() + "/" + pt.Namespace + "/" + pt.Type
+}
+
+// NewDefaultProvider returns the default address of a HashiCorp-maintained,
+// Registry-hosted provider.
+func NewDefaultProvider(name string) Provider {
+	return Provider{
+		Type:      name,
+		Namespace: "hashicorp",
+		Hostname:  "registry.terraform.io",
+	}
+}
+
+// NewLegacyProvider returns a mock address for a provider.
+// This will be removed when ProviderType is fully integrated.
+func NewLegacyProvider(name string) Provider {
+	return Provider{
+		Type:      name,
+		Namespace: "-",
+		Hostname:  "registry.terraform.io",
```

Comment:
I like the idea of squatting on a local convention that is equivalent to the one used in the provider registry API for "default" providers, cause that existing assumption blocks us from ever bestowing this `registry.terraform.io/-/foo` form with any other meaning that might conflict with this legacy fallback situation. :+1: